pennylane/qnodes/base.py

```diff
@@ -58,30 +59,36 @@ def _get_signature(func):
     Adds the following attributes to func:
 
     * :attr:`func.sig`: OrderedDict[str, SignatureParameter]: mapping from parameters' names to their descriptions
-    * :attr:`func.n_pos`: int: number of required positional arguments
-    * :attr:`func.var_pos`: bool: can take a variable number of positional arguments (``*args``)
-    * :attr:`func.var_keyword`: bool: can take a variable number of keyword arguments (``**kwargs``)
+    * :attr:`func.n_pos`: int: number of positional arguments
+    * :attr:`func.n_pos_nd`: int: number of positional arguments without default value
+    * :attr:`func.var_pos`: None, str: name of the var-positional argument (``*args``), or None if there isn't one
+    * :attr:`func.var_keyword`: None, str: name of the var-keyword argument (``**kwargs``), or None if there isn't one
 
     Args:
         func (callable): function to introspect
     """
     sig = inspect.signature(func)
     # count positional args, see if VAR_ args are present
     n_pos = 0
-    func.var_pos = False
-    func.var_keyword = False
+    n_pos_nd = 0  # count positional params without default value
+    func.var_pos = None
+    func.var_keyword = None
     for p in sig.parameters.values():
         if p.kind <= inspect.Parameter.POSITIONAL_OR_KEYWORD:
             n_pos += 1
+            if p.default is inspect.Parameter.empty:
+                # positional parameter with no default value
+                n_pos_nd += 1
         elif p.kind == inspect.Parameter.VAR_POSITIONAL:
-            func.var_pos = True
+            func.var_pos = p.name
         elif p.kind == inspect.Parameter.VAR_KEYWORD:
-            func.var_keyword = True
+            func.var_keyword = p.name
 
     func.sig = OrderedDict(
         [(p.name, SignatureParameter(idx, p)) for idx, p in enumerate(sig.parameters.values())]
     )
     func.n_pos = n_pos
+    func.n_pos_nd = n_pos_nd
```

Comment:
Perhaps this is a thought for another PR, but it might be better to store all this information inside of an attribute of `QNode`. The reason being is that we are here defining new attributes of a `callable` instance, which by default might not be expected to be defined.

Therefore, this function could return some sort of dictionary or so, that contains all the data.

Reply:
In Python you can add instance attributes to a function, since the functions are objects. You can use this e.g. to mimic C-style static function variables.

Reply:
Right! My only concern here was that developers would need to know and keep in mind that `_get_signature` is called inside of `__init__` and that upon instantiation, these attributes become accessible by having e.g. `qnode.func.n_pos_nd`. On the other hand, I agree that logically it better fits the `qfunc`.